Update/experimental select control

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Exposes some functions that are provided by the `useCombobox` hook from the `downshift` library and adds two extra properties:
`clearSearchOnSelect` - Clears the search field once the user selects an item
`keepMenuOpenOnSelect` - Keeps the dropdown open once a user selects an item.


### How to test the changes in this Pull Request:

1. Load this branch and run: `pnpm run build && pnpm run storybook`
2. Load the `Custom Render` component of the experimental select control
3. Type something in or hit the down arrow and select an item, the dropdown should stay open.
4. Load the `Custom render clear on select` component and do a search and select an item. Notice how your search cleared input cleared after selection.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Combine last 23 commits into one.

Update css, so dropdown shows correctly

Add extra callbacks and expose some downshift combobox functions

Add changelog

Update stories to reflect new props

Update stories label

Removed `clearOnSelect` prop and reverted some other changes

Fix missed props during rebase

Remove old storybook story and slight fix from rebase

Remove unneeded code

Update default get filtered functionality and add regex escaping

Some updates by testing a new onStateChange callback

Expose the onStateChange callback for more fine control ove the useSelect

Some minor adjustments

Expose stateReducer and minor adjustments from PR review

Simplify logic around isOpen handling with outside component

Remove unused import

Remove onStateChange prop

Address PR feedback

Some minor adjustments to the input handling

Update the way the stateReducer is called

Remove some of the unneeded logic

Remove onStateChange and replace it with onSelectedItemChange and onInputValueChange for simplicity

Fix lint error

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -103,20 +114,51 @@ export const SelectControl = ( {
-						setInputValue(
-							multiple ? '' : itemToString( selectedItem )
-						);
+						if ( clearSearchOnSelect ) {
```
I have mixed feelings of this, but unless `clearSearchOnSelect` is a very common pattern we anticipate on using, I'd prefer to let the consumer dictate that.

We could provide an `onStateChange` to the consumer and `setInputValue` to allow better control over what happens with this select control.

louwie17 replied:
I could remove this one again, I originally thought I needed this for the category dropdown, but given the window will stay open, clearing the search would seem very odd.
I could only see some benefit for clearing on `InputBlur` but not for any of the other options.
But given that is not required for us, I can remove this for now.

louwie17 replied:
Actually, this logic is handy for us on a multi select, when using it with keep open. As things are a little strange when a user search for a value and selects one, then it auto clears showing a different list (while still open).

I could still remove the `clearSearchOnSelect` option, but make it not clear for `multiple` selection. Maybe only not clear for multiple selection and `keepMenuOpenOnSelect`

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -103,20 +114,51 @@ export const SelectControl = ( {
+				case useCombobox.stateChangeTypes.ItemClick:
+					return {
+						...changes,
+						isOpen: keepMenuOpenOnSelect ? true : false,
```
Is the addition of `stateReducer` mostly for the sake of this prop change?  If so, in line with my earlier comment, I think the consumer could handle controlling special open state or rendering.

There's some level of prior art with this as `isOpen` can be passed to the child for menu render.

louwie17 replied:
Going this route felled like a more native approach to `downshift`. If I wanted to handle the `isOpen` manually and pass it to the menu, I would have to keep a create a separate state outside of the `SelectControl` component to toggle this, and make sure it adheres to all the above events as well.
This would mean that we probably want to expose a callback to the `stateReducer` change anyhow.

To me this felled overkill, and adding an option for this here felled better. I was following this example of Downshift -> https://www.downshift-js.com/use-combobox/#basic-multiple-selection with leaving the dropdown open.

This feature does feel like it could be more popular. I could still expose a callback for the state reducer changes, if that seems of benefit?
Or would you prefer to expose the entire `stateReducer` function? Although this would require the user to have more knowledge of `downshift`.

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -130,18 +172,20 @@ export const SelectControl = ( {
 			{ /* eslint-enable jsx-a11y/label-has-for */ }
 			<div className="woocommerce-experimental-select-control__combo-box-wrapper">
-				{ multiple && (
+				{ multiple && multipleSelection && (
```
Any particular purpose for not destructuring props from `multipleSelection`?

The logic around this becomes a little more confusing IMO by having multiple checks for properties that should inherently exist.

louwie17 replied:
No particular reason, I think previously I was running into some naming overlap, so not de-structuring prevented that. But I should be able to change that back.

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -70,31 +72,40 @@ export const SelectControl = ( {
 
+	let currentSelectedItem = null;
+	if ( ! multiple && selectedItems.length > 0 ) {
+		currentSelectedItem = selectedItems[ 0 ];
```
Heads up that I have multiple changes to how state selection is handled landing in https://github.com/woocommerce/woocommerce/pull/34510 related to accessibility.

joshuatf reviewed (commented):
Great work overall here, @louwie17!  Left a couple thoughts and questions.

I'd like to land https://github.com/woocommerce/woocommerce/pull/34510 first so we can rebase this one.  There are a number of changes related to how Downshift handles its internal state that are required for the accessibility features to work.

I like the additions in making this component more customizable.  I think for more opinionated changes, we can leave them to the consumer, but if you feel strongly there are features that all select controls would use, let's discuss.

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -196,9 +230,12 @@ function SelectControl< ItemType = DefaultItemType >( {
 				highlightedIndex,
 				getItemProps,
 				getMenuProps,
+				getToggleButtonProps,
```
Not opposed to exposing this, but just checking the use case for this.

louwie17 replied:
Given I had enabled the stayOpenOnSelect for the category dropdown when the user selected the **Add new "category"** field and the modal opened, the dropdown would stay open.
Exposing this function, allowed me to call this and have it close again.

joshuatf commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -176,25 +176,35 @@ export const CustomRender: React.FC = () => {
 		setSelected( [ ...selected, item ] );
 	};
 
+	const getFilteredItems = ( allItems: ItemType[], inputValue: string ) => {
```
We could optionally import the `defaultGetFilteredItems`, right?

joshuatf replied:
On second thought, I like this filter better than the existing default one since it matches anywhere in the string instead of just checking if it the string starts with some input.

Should we move this logic to the default function?  Can be done in a follow-up if you think it's outside the scope of this PR.

louwie17 replied:
Yeah I can do that for sure.

joshuatf commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -176,25 +176,35 @@ export const CustomRender: React.FC = () => {
 		setSelected( [ ...selected, item ] );
 	};
 
+	const getFilteredItems = ( allItems: ItemType[], inputValue: string ) => {
```
I think this needs to be updated to `DefaultItemType`.

```suggestion
	const getFilteredItems = ( allItems: DefaultItemType[], inputValue: string ) => {
```

louwie17 replied:
Good catch

joshuatf commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -176,25 +176,35 @@ export const CustomRender: React.FC = () => {
 	};
 
+	const getFilteredItems = ( allItems: ItemType[], inputValue: string ) => {
+		const re = new RegExp( inputValue, 'gi' );
```
I think the `inputValue` needs to be escaped unless we want to allow regex input 😄 

Can either use lodash's `escapeRegExp` or vanilla JS (preferable IMO so we can eventually remove lodash) - `string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');`

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -150,6 +158,32 @@ function SelectControl< ItemType = DefaultItemType >( {
 					break;
 			}
 		},
+		stateReducer: ( state, actionAndChanges ) => {
```
Instead of creating a new state reducer and `keepMenuOpenOnSelect` prop, could we introduce an `onBlur` and `onFocus` prop to the component?

This gives the consumer a lot more control over the menu interaction and is less opinionated about the state of openness.

We could add a new case to the above `onStateChange` reducer:

```
				case useCombobox.stateChangeTypes.InputBlur:
					onBlur();
```

And for the focus, we can add it to the combobox below:

```
						onFocus: () => {
							setIsFocused( true );
							onFocus();
						},
```

Then in the consumer (the story):

```
const [ isOpen, setIsOpen ] = useState( false );

<SelectControl
	onBlur={ () => setIsOpen( false ) }
	onFocus={ () => setIsOpen( true ) }
>
	<Menu isOpen={ isOpen } getMenuProps={ getMenuProps }>
</SelectControl>
```

louwie17 replied:
Yeah I could give this solution a try, one thing I do wonder though if it would cause a conflict between the `isOpen` state of combobox and the `isOpen` state of the outside component.
Currently the only item I can see relying on this is the `getDropdownProps` not sure how important that is. Unless we also allow the option for the user to pass in the `isOpen` state to the `SelectControl` component.
Let me see if I can wrangle something together.

joshuatf reviewed (commented):
Thanks for rebasing and sorry I wasn't able to get the previous accessibility PR merged sooner.  I know this PR has been sitting for a while.

I left some suggestions about how we handle open state that I think might be worth implementing.  Let me know what you think.

louwie17 commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -113,22 +132,53 @@ function SelectControl< ItemType = DefaultItemType >( {
+		initialIsOpen: isOpen,
+		defaultIsOpen: isOpen,
 		items: filteredItems,
+		selectedItem: null,
```
I had to add this, so the `selectedItem` is always passed whenever a item in the list is clicked, this allows us to un-select it right away if keep the selected items as part of the list.
I followed this example: https://www.downshift-js.com/use-combobox#basic-multiple-selection

louwie17 commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -39,13 +43,25 @@ type SelectControlProps< ItemType > = {
 	onInputChange?: ( value: string | undefined ) => void;
 	onRemove?: ( item: ItemType ) => void;
 	onSelect?: ( selected: ItemType ) => void;
+	onFocus?: ( data: { inputValue: string } ) => void;
```
Still needed `onFocus` as that is not triggered through the `onStateChange`.

joshuatf replied:
Very frustrating that they don't offer this in their API.  I'm definitely hoping we can make this available in our state changes once we remove Downshift.

louwie17 commented:
@joshuatf I made the changes as we discussed yesterday over video, let me know what you think.

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -113,22 +132,53 @@ function SelectControl< ItemType = DefaultItemType >( {
+				onStateChange &&
+				! onStateChange(
+					{ inputValue: value, type, selectedItem },
+					{
```
Any reason that we don't combine this object with the line above?  Seems like it might be easier for consumers to parse.

I'm guessing it's separated to match Downshift's API, but I think as far as consumers are concerned, we can dictate our own API.

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -102,9 +113,17 @@ function SelectControl< ItemType = DefaultItemType >( {
 
 	const {
-		isOpen,
+		isOpen: comboboxIsOpen,
```
Nice naming!  Better than `cIsOpen` 😆

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -113,22 +132,53 @@ function SelectControl< ItemType = DefaultItemType >( {
+			) {
+				return;
+			}
 			switch ( type ) {
```
I liked your original idea of passing in the parent state handler and letting the child call that since  I think it might be more obvious to consumers who the API is used.  Any reason for the change of heart?

E.g.,

```
<SelectControl
  onStateChange={ ( { defaultOnStateChange } ) => {
    ...
   defaultOnStateChange();
  } }
/>
```

louwie17 replied:
No particular reason, I think as I was doing it, it felled simpler to go with your approach. But that doesn't make it overly clear to the consumer.
passing in the defaultOnStateChange, isn't that much work.

joshuatf commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -162,30 +162,80 @@ export const Async: React.FC = () => {
 		setSelected( [ ...selected, item ] );
 	};
 
+	const getFilteredItems = (
```
Just confirming that we're moving this to the `SelectControl` in a follow up.  Like this filtering better than the currently implemented one  😄

louwie17 replied:
I did move it there actually, this one is just slightly different. It actually doesn't filter out the selected items that the default filtered items does.

joshuatf commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -162,30 +162,80 @@ export const Async: React.FC = () => {
+					switch ( type ) {
+						case selectControlStateChangeTypes.InputBlur:
+						case selectControlStateChangeTypes.InputKeyDownEscape:
+							setInputValue( '' );
```
I think this logic could be moved to the default state change handler.

joshuatf commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -162,30 +162,80 @@ export const Async: React.FC = () => {
+						case selectControlStateChangeTypes.InputKeyDownEscape:
+							setInputValue( '' );
+							setIsOpen( false );
+							setIsFocused( false );
```
I think we shouldn't remove focus on this when hitting escape.  See https://www.w3.org/WAI/ARIA/apg/patterns/combobox/#listbox-popup-keyboard-interaction

joshuatf commented on packages/js/components/src/experimental-select-control/stories/index.tsx:
```diff
@@ -162,30 +162,80 @@ export const Async: React.FC = () => {
+							break;
+						case selectControlStateChangeTypes.InputKeyDownEnter:
+						case selectControlStateChangeTypes.FunctionSelectItem:
+						case selectControlStateChangeTypes.ItemClick:
```
I probably should have looked this up earlier, but I think the correct action on selection is to close the list box and not keep it open.  https://www.w3.org/WAI/ARIA/apg/patterns/combobox/#listbox-popup-keyboard-interaction

I'm not sure if leaving the box open on click and only close it when using the keyboard satisfies the accessibility concerns.  But I don't want to hold you up on this PR and we can address later; there are other good changes worth merging in this PR.

joshuatf reviewed (commented):
Just left a few comments/questions, but nothing major.  Overall this is looking great!

louwie17 commented:
@joshuatf I just updated this with some of your suggested changes.
And as mentioned in Slack already, I did end up having to expose the `stateReducer` to handle the `highlightIndex`, which makes me wonder if we should only expose the `stateReducer` anyhow and forgo the use of exposing `onStateChange`.

If we did that, I wouldn't need to handle the `isOpen` state outside of the SelectControl and can leave the component still handling that. The `stateReducer` does feel like it would be simpler, and require less changes.

louwie17 commented:
@joshuatf I updated this, we now only need this logic to keep the dropdown open for clicks by the consumer:
```
stateReducer={ ( state, actionAndChanges ) => {
					const { changes, type } = actionAndChanges;
					switch ( type ) {
						case selectControlStateChangeTypes.ItemClick:
							return {
								...changes,
								isOpen: true,
								inputValue: state.inputValue,
								highlightedIndex: state.highlightedIndex,
							};
						default:
							return changes;
					}
				} }
```

I moved some of the more conditional logic out of the `onStateChange` function and moved it to the `stateReducer`, to more accurately reflect the changes. This way we can rely on the `inputValue` passed into the `onStateChange` to always be the one we want to update too. This allowed things to stay simple and allowed more flexibility for the `stateReducer`.

I currently left the `onStateChange` as an prop that allows users to pass in an updated `inputValue`. Although `stateReducer` can be used for this, so I could probably remove that (on second thought, I will probably remove it).

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -54,10 +67,10 @@ function SelectControl< ItemType = DefaultItemType >( {
 		getItemProps,
 		getMenuProps,
-		isOpen,
+		isOpen: comboboxIsOpen,
```
We can just keep this as `isOpen` for simplicity now, right?

louwie17 replied:
True, I fixed that as part of: https://github.com/woocommerce/woocommerce/pull/34620/commits/ef3a13a01571a4ece613944f6af96990b6407a33

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -113,42 +128,84 @@ function SelectControl< ItemType = DefaultItemType >( {
 		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
 		// @ts-ignore
-		onStateChange: ( { inputValue: value, type, selectedItem } ) => {
+		onStateChange: ( changes ) => {
```
Is it possible to move all (or at least most of) the logic in the `onStateChange` to the `stateReducer`? I think this would simply this component a good bit if we could just look to one area for changes.

louwie17 replied:
Unfortunately I don't think so, I purposely left all the function calls that calls our state handlers in the `onStateChange`, as it felled out of place to trigger other changes within the `stateReducer` function, I think that function is mostly meant to stay pure, without any side effects. Pretty well all of the items left in the `onStateChange` are side effects.

joshuatf replied:
I think as is, the state reducer is fairly broken.  Here are a couple of the issues I've run into:

In a multiple select control, hit enter on an item.  The input value gets updated to the selected item's text along with adding the item:
<img width="197" alt="Screen Shot 2022-09-29 at 2 44 47 PM" src="https://user-images.githubusercontent.com/10561050/193147924-2672b57a-8a8c-44cc-a32e-54f13c0dec2c.png">

In a single select control, select a value.  Afterwards remove a character or more and blur the input.  See that the input value gets removed despite still having a selected value.
<img width="332" alt="Screen Shot 2022-09-29 at 2 44 15 PM" src="https://user-images.githubusercontent.com/10561050/193147929-23af9715-b404-4afc-a6af-b5cf65b4ac3b.png">
<img width="296" alt="Screen Shot 2022-09-29 at 2 44 21 PM" src="https://user-images.githubusercontent.com/10561050/193147928-c6436a24-b4c4-4843-b2ed-c916341cd2d1.png">

joshuatf replied:
I would think that we could leave the `stateReducer` very simple and keep our side effects as they were in the `onStateChange`.  For example:

```js
		stateReducer: ( state, actionAndChanges ) => {
			const { changes } = actionAndChanges;
			const newChanges = stateReducer
				? stateReducer( state, actionAndChanges )
				: {};
			return {
				...changes,
				...newChanges,
			};
```

It seems like there's a lot of manipulation around the input value in the `stateReducer` in your current changes.  If your use case needs to be able to control the input value, maybe we should rely on Downshift's internal `inputValue` instead of handling our own.  That way, we can just have a `stateReducer` and won't need duplicate logic in the `onStateChange`.

>as it felled out of place to trigger other changes within the stateReducer function, I think that function is mostly meant to stay pure, without any side effects

I understand the reasoning here and this would be my preference as well, though I'm not sure it would cause any real problems if we need to manipulate our own state within this reducer.

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -113,42 +128,84 @@ function SelectControl< ItemType = DefaultItemType >( {
 						onSelect( selectedItem );
 						if ( multiple ) {
 							addSelectedItem( selectedItem );
-							setInputValue( '' );
```
I think the accessible interaction might be to clear the input value after selection.  Otherwise it may not be obvious to non-sighted users that they are continuing to add to an existing search string. If this is a custom interaction with the new component, maybe we could just keep the text as part of its state reducer for clicked items?

louwie17 replied:
Nice catch actually, the `setInputValue` was being called further down in the case statement, but I forgot that I left the `break;` in, which would never call it, I also updated the `stateReducer` to reflect this logic.
This should now work as intended, fixed as part of: https://github.com/woocommerce/woocommerce/pull/34620/commits/ef3a13a01571a4ece613944f6af96990b6407a33

joshuatf reviewed (commented):
Awesome work, @louwie17!  Looking almost perfect; just left a few minor clean-up comments.  Let me know if you have any questions 😄

louwie17 commented:
@joshuatf I addressed the changes as per our zoom discussion yesterday, let me know what you think.
I tested most scenarios and things seemed to work well, I don't have the screen reader stuff set up for testing though, so you may want to double check that.

louwie17 committed:
Remove duplicate and unused import

louwie17 committed:
Fix call to ControlledPropUpdatedSelectedItem in story

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -121,45 +134,56 @@ function SelectControl< ItemType = DefaultItemType >( {
+		onInputValueChange: ( changes ) => {
+			if ( changes.inputValue !== undefined ) {
+				setInputValue( changes.inputValue );
+				if ( changes.isOpen ) {
```
Just out of curiosity, we do not call `onInputChange` when the menu isn't open to avoid using this callback when resetting to the input value (to the selected value for single or blank when multiple)?

I don't have a strong opinion on this right now, but just wondering out loud whether or not we should always call this on change and let the consumer decide.  That might mean passing more information (`changes`) to the consumer though to let them decide.

Either way, I don't think it needs to hold up this PR and we can follow up on this.

louwie17 replied:
> Just out of curiosity, we do not call onInputChange when the menu isn't open to avoid using this callback when resetting to the input value (to the selected value for single or blank when multiple)?

correct.

> I don't have a strong opinion on this right now, but just wondering out loud whether or not we should always call this on change and let the consumer decide. That might mean passing more information (changes) to the consumer though to let them decide.

Yeah, that would work, passing back `changes` would make the most sense in that circumstance. My current interpretation was that our `onInputChange` callback only pertained to a user searching in the input field, which is why I stuck with that above logic. I don't mind changing this in the future though.

joshuatf replied:
>My current interpretation was that our onInputChange callback only pertained to a user searching in the input field, which is why I stuck with that above logic. 

Ya, that makes sense to me right now.  I'm not sure if we have a case that will need this callback fired under all circumstances, but we can leave it as is until that point comes.

joshuatf reviewed (dismissed):
Testing well!  Really appreciate all your hard efforts and discussion on this PR, Lourens! 💯 🎉 🚢

joshuatf commented on packages/js/components/src/experimental-select-control/select-control.tsx:
```diff
@@ -218,6 +245,10 @@ function SelectControl< ItemType = DefaultItemType >( {
 						getItemValue,
+						selectItem,
+						setInputValue,
+						openMenu,
```
We're not currently using these anymore, right?  Just checking- not opposed to leaving them in as they could be useful to a toggle button.

louwie17 replied:
I am not a 100% sure if we need `setInputValue` anymore, I will have to check the category code out again.
I will probably still need the `openMenu` and the `selectItem` functions.

joshuatf reviewed (approved):
One general question, but this is looking good and still testing well!  Thanks again, Lourens!  :shipit:
